File: internal/app/repo/postgres/package_repo.go
package postgres

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"mahir-trade-be/internal/app/repo/postgres/queries"

	"go.uber.org/dig"
)

type (
	PackageRepo interface {
		CreatePackage(ctx context.Context, req models.Package) (id int, err error)
		GetPackages(ctx context.Context, req models.GetPackagesRequest) (packages []models.Package, totalCount int64, err error)
		GetPackageByID(ctx context.Context, id int64) (res models.Package, err error)
		UpdatePackage(ctx context.Context, req models.Package) (err error)
		SoftDeletePackage(ctx context.Context, id int64, deletedBy string) (err error)
	}

	PackageRepoImpl struct {
		dig.In

		*sql.DB
	}
)

func NewPackageRepo(impl PackageRepoImpl) PackageRepo {
	return &impl
}

func (p *PackageRepoImpl) CreatePackage(ctx context.Context, req models.Package) (id int, err error) {
	rows, err := p.QueryContext(ctx, queries.QueryCreatePackage, req.Price, req.DurationInMonth, req.Description)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while CreatePackage err: %v", err.Error()))
		return id, err
	}

	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&id)
		if err != nil {
			slog.ErrorContext(ctx, fmt.Sprintf("error while CreatePackage err: %v", err.Error()))
			return id, err
		}
	}

	return id, nil
}

func (p *PackageRepoImpl) GetPackages(ctx context.Context, req models.GetPackagesRequest) (packages []models.Package, totalCount int64, err error) {
	if req.Limit == 0 {
		req.Limit = 10
	}

	offset := 0
	if req.Page > 1 {
		offset = int((req.Page - 1) * req.Limit)
	}

	rows, err := p.QueryContext(ctx, queries.QueryGetPackages, req.Limit, offset)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while GetPackages err: %v", err.Error()))
		return packages, totalCount, err
	}

	defer rows.Close()

	for rows.Next() {
		var pack models.Package
		err = rows.Scan(&pack.ID, &pack.Price, &pack.DurationInMonth, &pack.Description, &pack.CreatedBy, &pack.UpdatedBy, &pack.CreatedAt, &pack.UpdatedAt)
		if err != nil {
			slog.ErrorContext(ctx, fmt.Sprintf("error while GetPackages err: %v", err.Error()))
			return packages, totalCount, err
		}

		packages = append(packages, pack)
	}

	totalCountRow := p.QueryRowContext(ctx, queries.QueryGetTotalPackages)
	err = totalCountRow.Scan(&totalCount)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while GetTotalPackages err: %v", err.Error()))
		return packages, totalCount, err
	}

	return packages, totalCount, nil
}

func (p *PackageRepoImpl) GetPackageByID(ctx context.Context, id int64) (res models.Package, err error) {
	row := p.QueryRowContext(ctx, queries.QueryGetPackageByID, id)
	err = row.Scan(&res.ID, &res.Price, &res.DurationInMonth, &res.Description, &res.CreatedBy, &res.UpdatedBy, &res.CreatedAt, &res.UpdatedAt)
	if err != nil {
		if err == sql.ErrNoRows {
			return res, fmt.Errorf("package with id %d not found", id)
		}

		slog.ErrorContext(ctx, fmt.Sprintf("error while GetPackageByID err: %v", err.Error()))
		return res, err
	}

	return res, nil
}

func (p *PackageRepoImpl) UpdatePackage(ctx context.Context, req models.Package) (err error) {
	_, err = p.ExecContext(ctx, queries.QueryUpdatePackage, req.Price, req.DurationInMonth, req.Description, req.ID)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while UpdatePackage err: %v", err.Error()))
		return err
	}

	return nil
}

func (p *PackageRepoImpl) SoftDeletePackage(ctx context.Context, id int64, deletedBy string) (err error) {
	_, err = p.ExecContext(ctx, queries.QuertSoftDeletePackage, deletedBy, deletedBy, id)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while SoftDeletePackage err: %v", err.Error()))
		return err
	}

	return nil
}

File: src/repo/package_repo.ts
import { Pool } from "pg";

import { Package } from "../models/package.model";

import db from "../db/db.config";
import { PackageQueries } from "./queries/package_queries";
import { PaginationRequest } from "@utils/response";

export class PackageRepository {
  private pg: Pool;

  constructor() {
    this.pg = db;
  }

  // --- CREATE PACKAGE ---
  async createPackage(req: Package): Promise<number> {
    try {
      const result = await this.pg.query<{ id: number }>(
        PackageQueries.CreatePackage,
        [
          req.price,
          req.durationInMonth,
          req.description,
          req.discountedPrice,
          req.discountExpired,
          req.createdBy,
        ]
      );

      if (result.rowCount === 0) {
        throw new Error("failed to create package");
      }

      return result.rows[0].id;
    } catch (err: any) {
      console.error(`[packageRepo][createPackage] error: ${err.message}`);
      throw err;
    }
  }

  // --- GET PACKAGES ---
  async getPackages(
    req: PaginationRequest
  ): Promise<{ packages: Package[]; totalCount: number }> {
    try {
      if (req.limit === 0) {
        req.limit = 10;
      }

      let offset = 0;
      if (req.page > 1) {
        offset = (req.page - 1) * req.limit;
      }

      const result = await this.pg.query(PackageQueries.GetPackages, [
        req.limit,
        offset,
      ]);

      const packages: Package[] = (result.rows as any[]).map((row) => ({
        id: row.id,
        price: row.price,
        durationInMonth: row.duration_in_month, // snake_case dari DB
        description: row.description,
        discountedPrice: row.discounted_price,
        discountExpired: row.discount_expired,
        createdBy: row.created_by,
        updatedBy: row.updated_by,
        createdAt: row.created_at,
        updatedAt: row.updated_at,
      }));

      const totalCount =
        result.rows.length > 0 ? parseInt(result.rows[0].total_count, 10) : 0;

      return { packages, totalCount };
    } catch (err: any) {
      console.error(`[packageRepo][getPackages] error: ${err.message}`);
      throw err;
    }
  }

  // --- GET PACKAGE BY ID ---
  async getPackageByID(id: number): Promise<Package> {
    try {
      const result = await this.pg.query<Package>(
        PackageQueries.GetPackageByID,
        [id]
      );

      if (result.rowCount === 0) {
        throw new Error(`package with id ${id} not found`);
      }

      return result.rows[0];
    } catch (err: any) {
      console.error(`[packageRepo][getPackageByID] error: ${err.message}`);
      throw err;
    }
  }

  // --- UPDATE PACKAGE ---
  async updatePackage(req: Package): Promise<boolean> {
    try {
      const result = await this.pg.query(PackageQueries.UpdatePackage, [
        req.price,
        req.durationInMonth,
        req.description,
        req.discountedPrice,
        req.id,
        req.updatedBy,
      ]);

      return (result.rowCount ?? 0) > 0;
    } catch (err: any) {
      console.error(`[packageRepo][updatePackage] error: ${err.message}`);
      throw err;
    }
  }

  // --- SOFT DELETE PACKAGE ---
  async softDeletePackage(id: number, deletedBy: string): Promise<boolean> {
    try {
      const result = await this.pg.query(PackageQueries.SoftDeletePackage, [
        deletedBy,
        deletedBy,
        id,
      ]);

      return (result.rowCount ?? 0) > 0;
    } catch (err: any) {
      console.error(`[packageRepo][softDeletePackage] error: ${err.message}`);
      throw err;
    }
  }

  // --- UPDATE PACKAGE DISCOUNT EXPIRED ---
  async updatePackageDiscountExpired(time: Date): Promise<boolean> {
    try {
      const result = await this.pg.query(
        PackageQueries.UpdatePackageDiscountExpired,
        [time]
      );

      return (result.rowCount ?? 0) > 0;
    } catch (err: any) {
      console.error(
        `[packageRepo][updatePackageDiscountExpired] error: ${err.message}`
      );
      throw err;
    }
  }
}
